src/seccomp/src/compiler.rs

```diff
@@ -0,0 +1,561 @@
+                 conditions. Use the `syscall` property instead or remove the `args` property.",
+                syscalls
+            ),
+        }
+    }
+}
+
+/// Deserializable object encapsulating rules for one or multiple syscalls
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct SyscallObject {
+    /// Name of the syscall
+    syscall: Option<String>,
+    /// Vector of syscall names
+    syscalls: Option<Vec<String>>,
+    /// Optional on-match action
+    action: Option<SeccompAction>,
+    /// Rule conditions
+    #[serde(rename = "args")]
+    conditions: Option<Vec<SeccompCondition>>,
+    /// Unused field, represents a comment property in the JSON format
+    #[allow(dead_code)]
+    comment: Option<String>,
+}
+
+impl SyscallObject {
+    #[cfg(test)]
+    pub fn new(
+        syscall: Option<String>,
+        syscalls: Option<Vec<String>>,
+        action: Option<SeccompAction>,
+        conditions: Option<Vec<SeccompCondition>>,
+    ) -> SyscallObject {
+        SyscallObject {
+            syscall,
+            syscalls,
+            action,
+            conditions,
+            comment: None,
+        }
+    }
+
+    /// True if object refers to a single syscall
+    fn is_singular(&self) -> bool {
+        self.syscall.is_some() && self.syscalls.is_none()
+    }
+
+    /// True if object refers to multiple syscalls
+    fn is_plural(&self) -> bool {
+        self.syscall.is_none() && self.syscalls.is_some()
+    }
+
+    /// Perform semantic checks after deserialization
+    fn validate(&self) -> Result<()> {
+        if self.syscall.is_some() && self.syscalls.is_some() {
+            return Err(Error::MultipleSyscallFields);
+        }
+
+        if self.is_plural() && self.conditions.is_some() {
+            return Err(Error::SyscallsWithArgs(
+                // safe to unwrap because self.is_plural() == true
+                self.syscalls.as_ref().unwrap().clone(),
+            ));
+        }
+
+        // validate all `SeccompCondition`s
+        if self.conditions.is_some() {
+            return self
+                .conditions
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|cond| cond.validate())
+                .find(|result| result.is_err())
+                .or_else(|| Some(Ok(())))
+                .unwrap()
+                .map_err(Error::SeccompFilter);
+        }
+
+        Ok(())
+    }
+}
+
+/// Deserializable seccomp filter. Refers to one thread category.
+#[derive(Deserialize, PartialEq, Debug)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct Filter {
+    /// Default action if no rules match. e.g. `Kill` for an AllowList
+    default_action: SeccompAction,
+    /// Default action if a rule matches. e.g. `Allow` for an AllowList
+    filter_action: SeccompAction,
+    /// The collection of rules
+    filter: Vec<SyscallObject>,
```

Comment:
Why not call it `rules` then?

Reply:
makes sense

Reply:
on a second thought, it doesn't make sense. I will update the comment to smth like `The collection of SyscallObjects`.

At this stage they have not yet been transformed in rules

Reply:
done